src/Foundation/NSObject2.cs

```diff
@@ -475,6 +475,36 @@ public virtual bool ConformsToProtocol (NativeHandle protocol)
 			if (!Runtime.DynamicRegistrationSupported)
 				return false;
 
+			// the linker/trimmer will remove the following code if the dynamic registrar is removed from the app
+			var classHandle = ClassHandle;
+			bool new_map = false;
+			lock (protocol_cache) {
+#if NET
+				ref var map = ref CollectionsMarshal.GetValueRefOrAddDefault (protocol_cache, classHandle, out var exists);
+				if (!exists)
+					map = new ();
+				ref var result = ref CollectionsMarshal.GetValueRefOrAddDefault (map, protocol, out exists);
+				if (!exists)
+					result = DynamicConformsToProtocol (protocol);
+#else
+				if (!protocol_cache.TryGetValue (classHandle, out var map)) {
+					map = new ();
+					new_map = true;
+					protocol_cache.Add (classHandle, map);
+				}
+				if (new_map || !map.TryGetValue (protocol, out var result)) {
+					result = DynamicConformsToProtocol (protocol);
+					map.Add (protocol, result);
+				}
+#endif
+				return result;
+			}
+		}
+
+		static Dictionary<NativeHandle, Dictionary<NativeHandle, bool>> protocol_cache = new ();
```

Comment:
I think we'll want this in the NSObject.Initialize method, to avoid adding a static cctor method in this class.

Reply:
You mean `ObjCRuntime.Runtime.Initialize` ? not `NSObject.Initialize` right ?

```csharp
internal static NativeHandle Initialize ()
{
	return class_ptr;
}
```

`NSObject` already has a static .cctor :(

```csharp
private const string selConformsToProtocol = "conformsToProtocol:";
private const string selEncodeWithCoder = "encodeWithCoder:";
public static readonly Assembly PlatformAssembly = typeof(NSObject).get_Assembly();
[BindingImpl((BindingImplOptions)3)]
private static readonly NativeHandle class_ptr = ObjCRuntime.Class.GetHandle("NSObject");
```

and an even larger one for macOS 😞 

but I don't have any problem to move the new dictionary inside `Runtime` 😄

Reply:
> You mean ObjCRuntime.Runtime.Initialize ? not NSObject.Initialize right ?

NSObject.Initialize is called from ObjCRuntime.Runtime.Initialize, so it doesn't matter.

> `NSObject` already has a static .cctor :(

Yep, but no need to make it bigger :) Maybe one day we'll get around to removing it.